File: services/strava_worker/weather.py
"""Reanalysis weather for recorded efforts.

The same ERA5 archive scripts/fetch_effort_history.py uses, and deliberately so:
the calibration is fitted across every effort at once, so mixing a second
weather source for recent rides would put a systematic step in the middle of the
training data that no metric would attribute to its cause.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime, timedelta, timezone

ARCHIVE = "https://archive-api.open-meteo.com/v1/archive"

HOURLY = ",".join(
    [
        "temperature_2m",
        "relative_humidity_2m",
        "surface_pressure",
        "precipitation",
        "wind_speed_10m",
        "wind_direction_10m",
        "wind_gusts_10m",
    ]
)

# ERA5 trails real time. An effort newer than this has no reanalysis yet, which
# is why ingestion records efforts without weather and a later pass fills it in
# rather than the two happening together.
ARCHIVE_LAG_DAYS = 6


def ready_before() -> datetime:
    """Efforts before this instant can be joined to reanalysis."""
    return datetime.now(timezone.utc) - timedelta(days=ARCHIVE_LAG_DAYS)


def month_bounds(year: int, month: int) -> tuple[str, str]:
    start = datetime(year, month, 1, tzinfo=timezone.utc)
    end = (start + timedelta(days=32)).replace(day=1) - timedelta(days=1)
    cutoff = ready_before()
    if end > cutoff:
        end = cutoff
    return start.strftime("%Y-%m-%d"), end.strftime("%Y-%m-%d")


def fetch_cell_month(lat: float, lon: float, year: int, month: int) -> dict[str, dict]:
    """Hourly reanalysis for one cell-month, keyed by ISO hour.

    Fetched per cell-month rather than per effort. A month of rides on the same
    roads is one request instead of dozens.
    """
    start, end = month_bounds(year, month)
    if start > end:
        return {}

    query = (
        f"latitude={lat:.4f}&longitude={lon:.4f}"
        f"&start_date={start}&end_date={end}"
        f"&hourly={HOURLY}&wind_speed_unit=ms&timezone=UTC"
    )
    request = urllib.request.Request(
        f"{ARCHIVE}?{query}",
        headers={"Accept": "application/json", "User-Agent": "windchaser-worker/1.0"},
    )
    try:
        with urllib.request.urlopen(request, timeout=45) as response:
            payload = json.loads(response.read().decode())
    except (urllib.error.HTTPError, urllib.error.URLError) as exc:
        print(f"[weather] archive unavailable for {lat},{lon} {year}-{month}: {exc}")
        return {}

    hourly = payload.get("hourly") or {}
    times = hourly.get("time") or []

    out: dict[str, dict] = {}
    for i, stamp in enumerate(times):
        def at(key: str):
            series = hourly.get(key) or []
            return series[i] if i < len(series) else None

        wind = at("wind_speed_10m")
        direction = at("wind_direction_10m")
        if wind is None or direction is None:
            continue
        out[stamp] = {
            "wind_speed_ms": wind,
            "wind_from_deg": direction,
            "gust_ms": at("wind_gusts_10m"),
            "temperature_c": at("temperature_2m"),
            "humidity_pct": at("relative_humidity_2m"),
            "pressure_hpa": at("surface_pressure"),
            "precip_mm": at("precipitation"),
        }
    return out


def hour_key(start_date: str) -> str:
    """The archive's hourly key for an effort's start, in UTC."""
    stamp = datetime.fromisoformat(start_date.replace("Z", "+00:00"))
    return stamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:00")

File: apps/web/lib/server/weather.ts
import "server-only";

import type { ForecastCell } from "../types";

const OPEN_METEO = "https://api.open-meteo.com/v1/forecast";

/** Forecast grid cell size in degrees, matching the fixture pipeline. */
export const CELL_DEG = 0.1;

export function cellOf(lat: number, lon: number): [number, number] {
  return [
    Math.round(lat / CELL_DEG) * CELL_DEG,
    Math.round(lon / CELL_DEG) * CELL_DEG,
  ];
}

export function cellId(cell: [number, number]): string {
  return `${cell[0].toFixed(2)},${cell[1].toFixed(2)}`;
}

interface CacheEntry {
  cell: ForecastCell;
  fetchedAt: number;
}
const cache = new Map<string, CacheEntry>();
const TTL_MS = 30 * 60 * 1000;

const HOURLY = [
  "temperature_2m",
  "relative_humidity_2m",
  "surface_pressure",
  "precipitation",
  "precipitation_probability",
  "wind_speed_10m",
  "wind_direction_10m",
  "wind_gusts_10m",
].join(",");

/**
 * One request per grid cell, shared by every segment in it. This is the caching
 * policy from docs/architecture/COST_STRATEGY.md, and the reason forecasts are
 * keyed by cell rather than by segment.
 */
export async function fetchCell(cell: [number, number]): Promise<ForecastCell> {
  const id = cellId(cell);
  const hit = cache.get(id);
  if (hit && Date.now() - hit.fetchedAt < TTL_MS) return hit.cell;

  const query = new URLSearchParams({
    latitude: cell[0].toFixed(4),
    longitude: cell[1].toFixed(4),
    hourly: HOURLY,
    wind_speed_unit: "ms",
    forecast_days: "7",
    timezone: "auto",
  });

  const res = await fetch(`${OPEN_METEO}?${query}`, { cache: "no-store" });
  if (!res.ok) throw new Error(`Open-Meteo ${res.status} for cell ${id}`);
  const payload = await res.json();
  const h = payload.hourly;

  const forecast: ForecastCell = {
    cell_id: id,
    latitude: payload.latitude,
    longitude: payload.longitude,
    timezone: payload.timezone,
    utc_offset_seconds: payload.utc_offset_seconds ?? 0,
    issued_at: new Date().toISOString(),
    time: h.time,
    temperature_c: h.temperature_2m,
    humidity_pct: h.relative_humidity_2m,
    pressure_hpa: h.surface_pressure,
    precip_mm: h.precipitation,
    precip_prob: h.precipitation_probability,
    wind_speed_ms: h.wind_speed_10m,
    wind_from_deg: h.wind_direction_10m,
    gust_ms: h.wind_gusts_10m,
  };

  cache.set(id, { cell: forecast, fetchedAt: Date.now() });
  return forecast;
}

export async function fetchCells(
  cells: Array<[number, number]>,
): Promise<Record<string, ForecastCell>> {
  const unique = new Map<string, [number, number]>();
  for (const c of cells) unique.set(cellId(c), c);

  const entries = await Promise.all(
    [...unique.values()].map(async (c) => {
      const forecast = await fetchCell(c);
      return [forecast.cell_id, forecast] as const;
    }),
  );
  return Object.fromEntries(entries);
}
